src/hotspot/share/ci/ciArray.cpp

```diff
@@ -95,7 +95,8 @@ ciConstant ciArray::element_value_impl(BasicType elembt,
 ciConstant ciArray::element_value(int index) {
   BasicType elembt = element_basic_type();
   GUARDED_VM_ENTRY(
-    return element_value_impl(elembt, get_arrayOop(), index);
+    ciConstant value = element_value_impl(elembt, get_arrayOop(), index);
+    return check_constant_value_cache(index, value);
```

Comment:
If you check the cache first, you don't need to be in VM when doing that. As 

(As it is now, it doesn't help static/instance fields, but it can be improved. Something for a follow-up cleanup.)